pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,371 @@
+}
+
+func (t *Tracker) IncrementReceivedSamples(lbs []mimirpb.LabelAdapter, value float64, now time.Time) {
+	if t == nil {
+		return
+	}
+	t.updateCountersWithLabelAdapter(lbs, now.Unix(), 0, value, 0, nil, true)
+}
+
+func (t *Tracker) IncrementActiveSeriesFailure() {
+	if t == nil {
+		return
+	}
+	t.totalFailedActiveSeries.Add(1)
+}
+
+func (t *Tracker) updateCountersWithLabelAdapter(lbls []mimirpb.LabelAdapter, ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	extractValues := func() []string {
+		labelValues := make([]string, len(t.labels))
+		for idx, cal := range t.labels {
+			for _, l := range lbls {
+				if l.Name == cal {
+					labelValues[idx] = l.Value
+					break
+				}
+			}
+			if labelValues[idx] == "" {
+				labelValues[idx] = missingValue
+			}
+		}
+		return labelValues
+	}
+	t.updateCountersCommon(extractValues, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+}
+
+func (t *Tracker) updateCounters(lbls labels.Labels, ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	extractValues := func() []string {
+		labelValues := make([]string, len(t.labels))
+		for idx, cal := range t.labels {
+			labelValues[idx] = lbls.Get(cal)
+			if labelValues[idx] == "" {
+				labelValues[idx] = missingValue
+			}
+		}
+		return labelValues
+	}
+	t.updateCountersCommon(extractValues, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+}
+
+func (t *Tracker) updateCountersCommon(
+	extractValues func() []string,
+	ts int64,
+	activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64,
+	reason *string,
+	createIfDoesNotExist bool,
+) {
+	// Extract label values
+	labelValues := extractValues()
+
+	// Reuse buffer from pool for building the observation key
+	buf := bufferPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufferPool.Put(buf)
+
+	// Construct the observation key by joining label values
+	for i, value := range labelValues {
+		if i > 0 {
+			buf.WriteRune(sep)
+		}
+		buf.WriteString(value)
+	}
+
+	// Lock access to the observation map
+	t.observedMtx.Lock()
+	defer t.observedMtx.Unlock()
+
+	// Update observations and state
+	t.updateObservations(buf.Bytes(), ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+	t.updateState(ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement)
+}
+
+// updateObservations updates or creates a new observation in the 'observed' map.
+func (t *Tracker) updateObservations(key []byte, ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	if o, known := t.observed[string(key)]; known && o.lastUpdate != nil {
```

Comment:
How can `o.lastUpdate` be nil?

Reply:
it can't, after changed to int64 instead of pointer